Add support for exception handling (C++ 6/6)

futile opened the pull request:
This PR is part of a chain of PRs that adds C++-support to KLEE. This is the final (and 'main') PR of the chain, the other PRs are:

- [x] ~~#961  - Adds basic support for `memalign`~~ (dropped; see the PR for infos; new PR: #988)
- [x] ~~#962  - Adds support for `newlocale` and `freelocale` (e.g., required for `cout`).~~ (will be done in `uclibc`; see the PR for infos)
- [x] #963  - ~~Implements some~~ Enables lowering of atomic LLVM instructions.
- [x] #964  - Adds a hack to support landingpad instructions.
- [x] ~~#965  - Adds capabilities to KLEE and KLEE's build system to allow linking against a `libcxx`.~~ Implemented in #1056.
- [ ] This PR (#966) - Adds support for exception handling.

These PRs form a chain, therefore they need to be merged from the top (e.g., top one first, while this PR can only be merged last). If required, there is some room to change the merge order, but some hard dependencies exist.

The main PRs are the two last ones, those which add support for building against a `libcxx` and support for exception handling. However, while the others are smaller, they are still required in order to analyze programs built and linked against a `libcxx` and using exceptions.

---

This PR adds support for exception handling and unwinding to KLEE, also supporting nested exceptions. The new tests show the added capabilities. 

This is implemented by keeping track of a current stack of exception frames per state once unwinding begins, i.e., `__cxa_throw` is called. When unwinding is invoked in LLVM, control flow will transfer to the next landingpad block in the current call chain. A landingpad can be specified in LLVM by using `invoke` instead of `call`, which takes the target landingpad - where control flow should resume in case of an exception - as an extra argument. Different `eh_typeids` can be used to continue control flow at different blocks depending on the type of the exception that is being thrown (e.g., when using different `catch`-blocks in C++). 

When a new exception is thrown, KLEE keeps track of the `bottom` and `top` stack frame indices of this exception. If execution unwinds below the `bottomStackFrameIndex`, it is uncaught, and an error is raised. The `topStackFrameIndex` adds support for nested exceptions, e.g., when a new exception is thrown inside a `catch`-block. As long as this new exception is caught inside this `catch`-block, it is not an error. Therefore, nested `catch`-blocks need to be supported, as well as nested `try`-blocks. See the tests `nested.cpp` and `nested_fail.cpp` for examples of this.

This PR also adds tests for a lot more general C++ functionality.

kren1 commented:
Just a note that https://github.com/klee/klee-uclibc/pull/20 has been merged. So locale should be good now.

aaronpuchert commented:
Just out of curiosity, since you write that this change depends on #965: is it not possible to have exception support independent of the standard library? The functions required by `throw` and `try`/`catch` seem to be handled separately anyway.

corrodedHash commented:
Hey @MartinNowack, I integrated your changes from the libc++ PR into this PR. It works now, but the MSAN test runs into a timeout.
I don't know how to handle this ideally, my approaches would be either to increase the worker timeout of travis (it times out after 289 of 294 tests), or to disable the MSAN run for the libcxx/CXX tests. What is your opinion on this?

ccadar commented:
@futile You should disable MSan for the C++ tests -- @MartinNowack , I believe this is what we decided, no?

corrodedHash commented:
@ccadar This sounds like a good solution, but I cannot find the option in the libcxx PR, or in the testing framework in general. How do you switch off a certain test specifically for the MSAN build?

kren1 commented:
@corrodedHash  You add `REQUIRES: not-memsan`

have a look at 

https://github.com/klee/klee/blob/0afe438667c2173c458c5f45adfe99506a076c61/test/Feature/MemoryLimit.c

futile commented:
I restricted the tests to only run when 'uclibc' is available. Due to a bug in the detection of 'uclibc'-availability for tests, this will prevent the tests from running on _all_ platforms until, e.g., #1123 is merged.

jbuening commented:
@futile I just noticed that you use `REQUIRES: not-memsan` as suggested by @kren1 above. However, it should be `REQUIRES: not-msan` (see #1128). Right now, the corresponding tests are always disabled.

futile commented:
@jbuening thanks for the tip! I changed `not-memsan` to `not-msan` in all the tests, and added a comment saying that we're disabling msan because otherwise CI times out.

futile commented:
The travis failure seems to be due to failed homebrew setup during the mac build. Probably needs a re-run?

I checked that all C++ tests (including the `REQUIRES: uclibc` ones) now run and exhibit the expected results (pass/xfail).

ccadar commented:
@futile I restarted it now and checks passed.  Is this PR ready for review now?

futile commented:
@ccadar thanks! Sorry for the slight delay, this PR should be ready for review. Opening up the LLVM-IR of, e.g., `test/CXX/symex/libc++/{exception,simple_exception}.cpp` should make it easier to see how the different parts of exception handling work together.

futile commented:
Rebased onto master branch.

futile commented:
CI failure seems to be spurious due to a failed docker-pull.

corrodedHash commented:
I changed the llvm::errs output to klee_warnings, commented the ExceptionFrame struct, and added/improved the tests. I'm now working on the more semantical changes.

I am not quite clear on why you want the --exit-on-error removed. To my understanding, this terminates Klee with a non-zero return value, which lets llvm lit know that this test failed, which is a desired behavior for tests that are not expected to have an error.
It would also be useful in the tests which expects an error, but there it is a problem that the return value is non-zero, because this lets llvm lit think that this was unexpected.

MartinNowack commented:
@corrodedHash Thanks a lot for your changes!  I'll have a look.

The `--exit-on-error` is fragile in the sense that you're testing for a non-zero return value of KLEE and this is not distinguishable from other errors that might happen for different reasons. We won't be able to spot such failing test case. Therefore, explicitly check what you're aiming for. In addition, this automatically adds some documentation to help understanding the test case.

MartinNowack commented:
@corrodedHash Sorry for my delayed answer. Thanks a lot for your changes. Like this, the PR is already in a better shape.
Did you manage to fix the remaining test cases?
```
    KLEE :: CXX/symex/libc++/landingpad.cpp
    KLEE :: CXX/symex/libc++/rethrow.cpp
```
Currently, it looks like they are failing.
Thanks for your patience.

Looking forward to get this merged.

futile commented:
Hey @MartinNowack! Sorry for the long radio-silence, but we have nevertheless made progress on this. Unfortunately (for us) however, @corrodedHash will be starting his bachelor's thesis soon. Therefore I will pick up the remaining work, but it'll be a bit more difficult to make time for it.

Anyway, while the tests are still failing, we found (we think) a good solution for inheritance-based try/catch-dynamics. Since the logic is a little bit more involved, we will be relying on `libcxxabi` to resolve the logic for us. `libcxxabi` is the runtime that performs exactly this resolution when compiling with clang normally (g++ also has a version of this). An easier option would have been if clang++/llvm offered this as library functionality, but I couldn't find an existing function in the llvm codebase that performs this resolution at the level we need, therefore my choice fell on using `libcxxabi`.

We already implemented a basic prototype that uses the logic in `libcxxabi` for the resolution, and it successfully solved an inheritance-based try/catch-scenario similar to the one in the test. What remains is properly integrating this into KLEE, which I hope I will get around to soon. Thanks for the continuous feedback!

arrowd commented:
In fact, FreeBSD uses `libcxxrt` as C++ runtime library, despite having clang and libc++ in its base. That `libcxxrt` looks smaller than `libcxxabi`, so take a look if you are interested.

MartinNowack commented:
@futile Great! I'm glad to hear about your progress. Those are excellent news. In case you push some updates, please ping me if possible. Sometimes, I don't get any notifications on changes of already opened PRs. Keep up the great work.

futile commented:
I just pushed an updated version of C++ exception handling that implements most of the missing features (known to me).

ping @MartinNowack :)

# Major Updates in this PR

## We now use `libcxxabi` to resolve whether a clause can catch an exception.

This properly handles inheritance and other things like catching arrays as pointers etc.
However, this requires a `libcxxabi` to be available for KLEE's build.
I added a CMake-option for this, but this will also require additional support in the CI/build-process.
Since `libcxxabi` is already built as part of building `libcxx`, however, this will probably be able to use that version.

I took a look at the CI-scripts, but if possible maybe somebody with better knowledge than me can add the required changes to these; I'll be glad to help out wherever necessary!

## Calling of Exception Destructors at the End of a `catch`-Block

The destructor function-pointer that is given to `__cxa_throw()` will now be called at the end of the `catch`-Block that handled the exception (more accurately, when `__cxa_end_catch()` is called for the exception).

## Exception Rethrowing

Code that rethrows an exception (using `throw;` in C++) now works correctly, as well as code that compiled down to a rethrow.

## Improved Handling of Nested Exceptions and Nested Throws

Nested exceptions, that is, throwing and catching an exception inside a `try`-block, have been improved.
There were some cases where these did not work correctly, and this PR fixes those.

## Implemented landingpad filter-clauses

This adds support for C++'s `throw()` specifications that limit which exceptions a function is allowed to throw.
Although this feature is not used much in practice, has been deprecated in C++11 and is removed in C++17, it was straight-forward to implement and can of course be useful for testing.

## Tests for all New Functionality

The tests should now cover all of the functionality implemented by this PR.
If there is something missing, please let me know/post a testcase, and I'll add it to this PR.

# Implementation Notes

For review (and documentation) purposes:

## Exceptions are kept on a per-`ExecutionState` stack

The `std::vector` `exceptionStack` keeps track of `n` currently caught exceptions (i.e., inside a `catch`-block but not yet fully handled), and of up to one uncaught/in-flight exception as its last element, which is currently unwinding.
If an in-flight exception leaves the stack frame of an exception being handled, this is reported as an error, since this will lead to a termination of the program.

## Implementation of C++'s exception-related functions in the `SpecialFunctionHandler`

The `SpecialFunctionHandler` now has handlers for functions related to C++'s exception handling, e.g., `__cxa_throw()`, `__cxa_begin_catch()`, `__cxa_rethrow()`, etc.

## Implementation of stack unwinding in `Executor::unwindToLandingpad`

This function implements stack unwinding.
If, during this process, an exception unwinds too far, the state will be terminated with a corresponding error.

## Landingpads handled by new runtime function: `_klee_eh_landingpad_handler`

Because handling landingpads includes determining whether a clause can catch an in-flight exception (taking, e.g., inheritance into account), we call into `libcxxabi` for this, and therefore also implemented the landingpad handler as a runtime function.
This function can be found in the new `klee-cxxabi` runtime.
To call this function, when we execute a landingpad instruction in the `Executor`, we serialize the information of this landingpad (see `Executor::serializeLandingpad`) and redirect control flow to `_klee_eh_landingpad_handler`.
In order for returning from this function to work, we add a special case to the handling of `Instruction::Ret` inside `Executor::executeInstruction`, which evaluates the return value of the handler function, and either continues or stops the stack unwinding.

# Known Missing Features

This section includes features I know about that are not implemented, but that are not essential for exception handling for most cases:

* Some exception handling related `__cxa_`-functions are not implemented (see the first link in the next section for a list of such functions)
* Support for setting/calling `std::unexpected_handler`, which has also been deprecated in C++11 and will be removed in C++17. 

# Useful Links

These are two links that proved to be especially helpful for understanding how C++'s exception handling works in LLVM IR:

* The section on "Exception Handling" of https://libcxxabi.llvm.org/spec.html
* LLVM's exception handling documentation under https://llvm.org/docs/ExceptionHandling.html#throw


Test/ask away! :)

ccadar commented:
Hi @futile, thanks again for this contribution!  Really sorry for getting behind with PR reviews, but I've discussed this with @MartinNowack and this is one of the next PR in line for review.

futile commented:
Hey @ccadar! Thanks! I just pushed a change that majorly overhauls our exception handling, and was just about to write up what changed/how it works now, so that should again help with review!

futile commented:
The version of exception handling I just pushed handles exceptions at the base-API level of the [Itanium  ABI](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html), instead of at the C++-level (`libcxxabi`). This means that C++-specific parts of exception handling that we implemented in previous versions are now handled by `libcxxabi`, e.g., keeping track of the stack of exceptions or aborting the process when an uncaught exception was found.

The reason for this is that `libcxxabi`'s exception handling is actually mostly regular code that is independent of the unwinding process itself, which is usually tightly coupled with the OS/platform.
For the unwinding process, `libcxxabi` uses `libunwind`'s `_Unwind_RaiseException`, which performs the actual unwinding process. We instead now implement `_Unwind_RaiseException` (and other exception-handling related functions from `libunwind`), which means that the "only" task KLEE has to perform is language-independent unwinding according to the Itanium ABI's unwinding base-API. As the personality function for exception handling is, however, also implemented at the binary-level, we implement a version of this for KLEE as well (this was already present in previous versions as the function that handles landingpads).

## High-Level Overview
The [Itanium ABI](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html) describes a two-phase unwinding process. In phase 1, the search phase, `_Unwind_RaiseException` has to check whether there exists a handler that will handle the exception being thrown. We implement this by iterating through the current stack (from top to bottom), and searching for invoke-instructions.

When we find an invoke instruction, we serialize the landingpad it points to into a simple format, and redirect execution to a call to our personality function (`_klee_eh_cxx_personality`). This function then uses `libcxxabi`-functionality to check whether any of the clauses or filter-clauses in the serialized landingpad clauses match the thrown exception (this was also done in the previous version, and mostly stayed the same).

We repeat this process until we either find a handling landingpad for the exception, or until we hit the bottom of the stack. In case we hit the bottom (i.e., no handler handles the exception), we return from `_Uwind_RaiseException` with an error, according to the Itanium ABI specification.

If we find a handler in phase 1, we switch to phase 2 of the unwinding process, the cleanup phase. In this phase the actual stack unwinding happens, until we reach the landingpad that we found in phase 1. However, while phase 1 skipped cleanup landingpads (as these don't count as "handling" the exception), phase 2 will execute all landingpads along the stack, including cleanup landingpads. This will, e.g., lead to correct execution of destructors.

After such a cleanup clause, the IR will eventually contain a `resume`-instruction, at which point we will continue the unwinding process. Once we reach the handling landingpad that we determined in phase 1 we will stop the unwinding process. In this phase, it is not necessary to call the personality function anymore, as we already know where unwinding will stop and what the selector value returned by the personality function will be (as we found these in phase 1).

By doing so, we have implemented the base-API part of exception handling according to the Itanium ABI. We leave the C++-specific parts to `libcxxabi`, which will be linked in as part of the specified `libcxx`.

## Implementation Notes

We shortly describe how the unwiding process is implemented in KLEE.

### Unwinding Start

Unwinding starts when the program calls the `_Unwind_RaiseException` function, which we implement in `runtime/klee-eh-cxx.cpp`. This function is a short wrapper calling `_klee_eh_Unwind_RaiseException_impl`, which is implemented as a SpecialFunction and starts the actual unwinding process.

### Phase 1/Search Phase

To start the search phase of the unwinding process, we store an instance of `SearchPhaseUnwindingInformation` in the state's `unwindingInformation` member (added in this PR) and call `Executor::unwindToNextLandingpad`. This function is called both in phase 1 and phase 2, and performs the actions corresponding to the currently active phase.

In phase 1, `unwindToNextLandingpad` will search the next landingpad in the stack that is not simply a cleanup-landingpad. In the search phase, the stack will not actually be unwound/changed, only searched. Upon finding a landingpad that potentially handles the exception, the function pushes a frame on top of the execution stack that calls into the personality function, `_klee_eh_cxx_personality`, and then continues execution of the state. In the executor, we then handle returns from the personality function (as part of the `Instruction::Ret`-handling), by either continuing the unwind-process if the landingpad doesn't catch the exception, or by switching to phase 2 if it does.

### Phase 2/Cleanup Phase

To enter the second phase, we store an instance of `CleanupPhaseUnwindingInformation` in the state's `unwindingInformation` member, to mark that the state is currently in the cleanup phase of unwinding. This instance contains the results of the search phase, such as the stack-frame at which the exception will be caught, as well as the selector-value returned by the personality function for this stack frame. Then we call `Executor::unwindToNextLandingpad` again, which, in phase 2, actually unwinds the stack by popping stack frames until a landingpad is reached, and then transfers control flow there.

We implemented the landingpad-instruction in `Executor::executeInstruction`. For all landingpads before the one that will catch the exception (which we know as a result of phase 1), we will return a `0` from the landingpad (together with the exception pointer), which will enter the cleanup-clause of the landingpad, if any. The IR will then eventually call the `resume`-instruction, which will cause unwinding to proceed to the next landingpad, until finally the catching landingpad is reached and we clean up the `unwindingInformation`, ending the unwinding process.

### `unwindingInformation` Storage

We store unwinding information inside the (new) `ExecutionState::unwindingInformation` field. This field contains an `std::unique_ptr`, so when no information is stored (i.e., unwinding is not currently happening), memory usage will be minimal. Only when unwinding is in progress will we require additional memory (though not a lot) to store the current unwinding state.

In addition to this PR and the Itanium ABI spec, taking a look at `libcxxabi`, especially `cxa_exception.cpp` and `cxa_personality.cpp` can give additional context for the changes, and the [LLVM page on exception handling](https://llvm.org/docs/ExceptionHandling.html) also gives additional insight.

## Testing

All KLEE-Testcases pass. I additionally also ran the tests contained in `libcxxabi` that are applicable to our implementation (no threading, exception handling turned on), to check the implementation. Of the 49 test cases, 44 passed and 5 failed. The failing test cases were mostly (4/5) checking unrelated functionality (e.g., backtrace generation) which either failed due to build system issues or because we don't support backtraces. The remaining failing test is related to exception handling, but failed in a call to `sleep`, and there in an external call to `__syscall_rt_sigaction`. The exception-handling part of the test works correctly, and it seems like it is `sleep` that doesn't work correctly (does `sleep` normally work in programs executed by KLEE?).

I also checked these tests against the previous version of our implementation, and there quite a few more of the tests failed. Some of them were because the tests use `std::set_unexpected`/`std::set_terminate` for checking, which was not supported before (but is now supported naturally by modeling unwinding at a lower level). Others looked like legitimate failures as we didn't implement all of the `__cxa`-functions before. The new implementation leaves all these to `libcxxabi`, and therefore doesn't reimplement its functionality, but implements only the unwinding process (and the personality function). I think that the new implementation fits in much better and more naturally with `libcxxabi`, and doesn't try to reimplement it anymore.

Overall - while I successfully used the previous implementation to find bugs before - I think the new version covers more cases, probably also fixes a few bugs, and just fits in cleaner.

futile commented:
The FreeBSD tests seem to fail as I'm using `isa_and_nonnull`, which only seems to be in LLVM since ~2019. These should be easy to replace.

The Travis tests seem to fail because KLEE's CMake now expects the `-DKLEE_LIBCXXABI_DIR` parameter, which should point to the directory in which `libcxxabi` exists (i.e., `llvm-project/libcxxabi`). I'd be happy if someone with more knowledge of KLEE's ci-build-system could point in the right direction on where these need to be added.

aaronpuchert commented on runtime/Makefile.cmake.bitcode:
```diff
@@ -17,4 +17,8 @@ ifneq ($(ENABLE_POSIX_RUNTIME),0)
 	DIRS += POSIX
 endif
 
+ifneq ($(BUILD_KLEE_EH_CXX),0)
```
You need to add this to `Makefile.cmake.bitcode.config.in`, otherwise it will just evaluate to "".

futile replied:
Hi, thanks for the comment! I have a set of patches by @jbuening lying around (thanks!), which fix this issue and improve build support in general, I'll try to make some time to push them soon.

futile replied:
Pushed them.

ccadar commented:
@futile I wanted to ping you about this PR -- like @MartinNowack I'm excited to see it so close to being merged!  Let me know if you need any assistance with the last bits pointed out by Martin.

jbuening commented:
@ccadar @MartinNowack I think I have addressed all open requests in this PR, hopefully bringing it further to being merged:
- CI should now pass with all LLVM versions. However, instead of taking the approach suggested by @ccadar for the test cases, I introduced a CMake variable `ENABLE_KLEE_EH_CXX` that can only be enabled for LLVM >= 8 and requires `ENABLE_KLEE_LIBCXX` to be true. Based on this, there is now a `eh-cxx` feature in `lit.cfg` which is used as a requirement for all tests with exception handling.
The CMake flag also allows to build KLEE with support for libcxx but without exception handling support. Finally, I adjusted the build scripts to enable exception handling for LLVM >= 8 if libcxx is enabled.
- An additional commit documents all C++ related flags (only some of which were introduced in this PR) in `README-CMake.md`.
- The remaining comments by @MartinNowack should all be addressed now. The only deviation I'm aware of is that instead of turning the `assert` into a `terminateStateOnExecError`, I replaced the latter into the former in the handler a few lines above (to make it consistent with the rest of `SpecialFunctionHandler` which uses `assert`s in virtually all cases).
- I ran `clang-format` on the patches and got rid of unneeded newlines and other formatting stuff (as noted by @MartinNowack) in the original commits.
- This PR was rebased on current master.

If there is anything else you would like me to do (that hopefully does not require a dive deep into the functionality of this PR), please let me know.

jbuening commented:
CI does indeed pass now. However, since `codecov/patch` is based on LLVM 6, and virtually everything in this PR is only enabled for LLVM >= 8, this check fails.

@ccadar @MartinNowack How do we proceed? Should I try to bump the LLVM version for codecov to 8/9/10?

ccadar commented:
@jbuening Fantastic to see this patch almost complete!
Yes, please try to update to LLVM 9.

jbuening commented:
@ccadar I changed the LLVM version to 9.0 in the coverage build. However, building that LLVM version (with `ENABLE_OPTIMIZED=0`) times out at 60 % (see https://travis-ci.com/github/klee/klee/jobs/381611101).

ccadar commented:
@jbuening I see you also triggered another run?  That's useful, but I see this one is already running for 1h20', so it will likely time out too.

The old coverage target on Travis CI was taking only ~15'.  I don't think the LLVM version is responsible for the huge time difference.  Is now libcxx also compiled with coverage info?

kren1 commented:
I would assume it's building the docker images. I think martin builds them locally and pushes to docker so that they don't get built in CI every time

ccadar commented:
@kren1 I see, I wasn't aware of that. 
@MartinNowack can you do that?  (In fact, we should switch all runs to LLVM 9).
Thanks!

linux40 commented:
I'm trying this patch to see if it could find some bugs caused by throwing of exceptions. I tried following example with llvm 8.0.0.
```cpp
struct bar {                      
  explicit bar(int i) { throw i; }
};                                
                                  
void foo() {                      
  int *p = new int;               
  bar{*p};                        
  delete p;                       
}                                 
                                  
int main() {                      
  try {                           
    foo();                        
  } catch (...) {                 
  }                               
}                                 
```
The above example contains a memory leak as `delete p;` will never be execute. I compiled the file with `clang++ -c -emit-llvm klee.cc` and ran `klee klee.bc` only to get a memory error of out-of-boud pointer. It seems that klee considers the return value of `__cxa_allocate_exception` as a null pointer as `test000001.ptr.err`, `warning.txt` and `assembly.ll` seem to suggest that. If so, this patch is far from practice.
```
Error: memory error: out of bound pointer
Stack:
        #000000050 in _ZN3barC2Ei (this=94535439160960, i=2880154539)
        #100000022 in _Z3foov ()
        #200000068 in main ()
Info:
        address: 94535439900720
        next: object at 94535439150848 of size 8
                MO12[8] allocated at _Z3foov():  %p = alloca i32*, align 8
```
```
KLEE: WARNING: undefined reference to variable: _ZTIi
KLEE: WARNING: undefined reference to function: __cxa_allocate_exception
KLEE: WARNING: undefined reference to function: __cxa_begin_catch
KLEE: WARNING: undefined reference to function: __cxa_end_catch
KLEE: WARNING: undefined reference to function: __cxa_throw
KLEE: WARNING: undefined reference to function: __gxx_personality_v0
KLEE: WARNING ONCE: Alignment of memory from call "_Znwm" is not modelled. Using alignment of 8.
KLEE: WARNING ONCE: calling external: __cxa_allocate_exception(4) at [no debug info]
```
```
; Function Attrs: noinline optnone uwtable
define linkonce_odr dso_local void @_ZN3barC2Ei(%struct.bar* %this, i32 %i) unnamed_addr #0 comdat align 2 {
entry:
  %this.addr = alloca %struct.bar*, align 8
  %i.addr = alloca i32, align 4
  store %struct.bar* %this, %struct.bar** %this.addr, align 8
  store i32 %i, i32* %i.addr, align 4
  %this1 = load %struct.bar*, %struct.bar** %this.addr, align 8
  %exception = call i8* @__cxa_allocate_exception(i64 4) #6
  %0 = bitcast i8* %exception to i32*
  %1 = load i32, i32* %i.addr, align 4
  store i32 %1, i32* %0, align 16
  call void @__cxa_throw(i8* %exception, i8* bitcast (i8** @_ZTIi to i8*), i8* null) #7
  unreachable

return:                                           ; No predecessors!
  ret void
}
```

kren1 commented:
You should run with `klee -libc=uclibc -libcxx -posix-runtime`

linux40 commented:
@kren1 Thank you, I will try this later.

linux40 commented:
I'd like to compile klee with libcxx support but stick with error `libc++ library not found at xxx` while `xxx` do contains `libc++.so` and many other relevant things. I checked the `CMakeError.log`, and found following messages:
```
...
fatal error: util.h: No such file or directory
...
fatal error: libutil.h: No such file or directory
...
undefined reference to `openpty'
...
```
But `/usr/include/libutil.h` do exist and I have specified `CMAKE_EXE_LINKER_FLAGS` with `libutil-freebsd.a`. Any help? I do not want to follow the way [offical document](https://klee.github.io/build-llvm9) shows, as it downloads another llvm rather than the one previously built.

kren1 commented:
Not sure why you don't want to follow the offical document, but you need an LLVM bitcode version of libc++.so. You can try compiling it yourself via `wllvm` and use the script as a guidance. You could also try the docker image, that should have libcxx built already.

linux40 commented:
Sorry for bothering so much. This time klee stops to report out of bound pointers, however it reports uncaught exception of type int while from my understanding, klee should report nothing. Maybe following warnings are relevant:
```
KLEE: WARNING: undefined reference to function: _Unwind_DeleteException
KLEE: WARNING: undefined reference to function: _Unwind_GetIP
KLEE: WARNING: undefined reference to function: _Unwind_GetLanguageSpecificData
KLEE: WARNING: undefined reference to function: _Unwind_GetRegionStart
KLEE: WARNING: undefined reference to function: _Unwind_RaiseException
KLEE: WARNING: undefined reference to function: _Unwind_SetGR
KLEE: WARNING: undefined reference to function: _Unwind_SetIP
KLEE: WARNING ONCE: Using zero size array fix for landingpad instruction filter
KLEE: WARNING ONCE: calling external: syscall(16, 0, 21505, 94582203603728) at src/runtime/POSIX/fd.c:1007 10
KLEE: WARNING ONCE: Alignment of memory from call "malloc" is not modelled. Using alignment of 8.
KLEE: WARNING ONCE: calling __klee_posix_wrapped_main with extra arguments.
KLEE: WARNING ONCE: Alignment of memory from call "_Znwm" is not modelled. Using alignment of 8.
KLEE: WARNING ONCE: calling external: _Unwind_RaiseException(94582205052768) at [no debug info]
terminating with uncaught exception of type int
KLEE: ERROR: (location information missing) abort failure
KLEE: NOTE: now ignoring this error at this location

KLEE: done: total instructions = 28835
KLEE: done: completed paths = 1
KLEE: done: generated tests = 1
```
And my example is now as follows:
```cpp
struct bar {                      
  explicit bar(int i) { throw i; }
};                                
                                  
void foo() {                      
  int *p = new int;               
  bar{*p};                        
  delete p;                       
}                                 
                                  
void baz() {                      
  try {                           
    foo();                        
  } catch (...) {                 
  }                               
}                                 
                                  
int main() { baz(); }             
```

kren1 commented:
You are still having external calls and undefined functions. Are you compiling KLEE with exception handling support? I think you need -DENABLE_KLEE_EH_CXX=ON to enable this PR also make sure cmake can find the libcxxabi source

aaronpuchert commented:
Just to clarify: this only requires `libc++abi`, not `libc++`, right?

kren1 commented:
Not sure, `__cxa_allocate_exception`&co might be in libc++

aaronpuchert commented:
My understanding is that `libc++`, like `libstdc++`, covers basically the `std` namespace, whereas language features like `throw` and `catch` are implemented by `libc++abi`. But I might be wrong.

kren1 commented:
Sure, but from the KLEE outputs @linux40 posted, it seems `__cxa_allocate_exception ` got defined after including libcxx. But that has a lot of assumption of what was actually run, so I might be wrong.

linux40 commented:
Thank you, klee now works as expected on this example.

jbuening commented:
@ccadar @MartinNowack after you merged #1329, I rebased this one and removed the commit that was testing to change just codecov to LLVM 9. So now everything is passing and while CI times have gone up a bit again (compared to #1329) they still look acceptable in my opinion.

MartinNowack reviewed (approved):
@futile Let's merge this - finally ;). I absolutely applaud you for adding this huge feature to KLEE.
I've been "just" reviewing this patches and it took me a lot of time to understand the full glory (aka pain) of exception handling details in C++. So, I can just imagine how much time it must have taken to implement this is in the first place. It's great that you contribute this to the KLEE project and is absolutely appreciated. I know, it took many iterations, but I think, the resulting implementation we've got is amazing!

Thank you so much for your really hard work!

@jbuening Of course, thank you so much as well for finalizing this PR.

Let's merge this.

ccadar commented:
Fully agreed, big thanks, @futile and @jbuening!!  And many thanks @MartinNowack as well, your review was critical in improving this patch!
